cmd/sqlflow/model_zoo_client.go

```diff
@@ -0,0 +1,205 @@
+	}
+	defer os.Chdir(curDir)
+	os.Chdir(repoDir)
+	err = tar.ZipDir(".", tarName)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(repoDir, tarName), nil
+}
+
+func releaseModel(opts *options) error {
+	if opts.DataSource == "" {
+		return fmt.Errorf("You should specify a datasource with -d")
+	}
+	db, err := database.OpenDB(opts.DataSource)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+	buf, err := model.LoadToBuffer(db, opts.ModelName)
+	if err != nil {
+		return err
+	}
+	modelData := buf.Bytes()
+	// (TODO:lhw) decode the model from the buffer when
+	// more metadata is stored in database, client should
+	// not parse the original model.TrainSelect directly,
+	// because parser should exist only on SQLFlow server
+
+	conn, err := getModelZooServerConn(opts)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	client := pb.NewModelZooServerClient(conn)
+	stream, err := client.ReleaseModel(context.Background())
+	if err != nil {
+		return err
+	}
+	nameParts := strings.Split(opts.ModelName, ".")
+	request := &pb.ReleaseModelRequest{
+		Name: nameParts[len(nameParts)-1],
+		Tag:  opts.Version,
+		// (TODO: lhw) add following fields form model
+		Description:       "",
+		EvaluationMetrics: "",
+		ModelClassName:    "MyDNNClassifier",
+		// (FIXME: lhw) should extract from the model metadata once
+		// we store it in the database
+		ModelRepoImageUrl: "test/my_repo:v1.0",
+		ContentTar:        modelData,
+		ContentUrl:        "",
+	}
+	stream.Send(request)
+	resp, err := stream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+	if !resp.Success {
+		return fmt.Errorf(resp.Message)
+	}
+	return nil
+}
+
+func deleteModel(opts *options) error {
+	conn, err := getModelZooServerConn(opts)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	client := pb.NewModelZooServerClient(conn)
+	// if user give a db.table format, we just use the table name
+	nameParts := strings.Split(opts.ModelName, ".")
+	req := &pb.ReleaseModelRequest{
+		Name: nameParts[len(nameParts)-1],
+		Tag:  opts.Version,
+	}
+	resp, err := client.DropModel(context.Background(), req)
+	if err != nil {
+		return err
+	}
+	if !resp.Success {
+		return fmt.Errorf(resp.Message)
+	}
+	return nil
+}
+
+func checkModelZooParam(opts *options) error {
+	// model zoo server is an optional param, so check it
+	if opts.ModelZooServer == "" {
+		return fmt.Errorf("you need to specify a model zoo server")
+	}
+	return nil
+}
+
+func releaseRepo(opts *options) error {
+	if err := checkModelZooParam(opts); err != nil {
+		return err
+	}
+	tarFile, err := tarRepo(opts.RepoDir,
+		filepath.Base(opts.RepoName)+".tar.gz")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tarFile)
+
+	conn, err := getModelZooServerConn(opts)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	client := pb.NewModelZooServerClient(conn)
+	stream, err := client.ReleaseModelRepo(context.Background())
+	if err != nil {
+		return err
+	}
+	buf, err := ioutil.ReadFile(tarFile)
```

Comment:
Try not to read and send buffer at once, the buffer maybe large.

Reply:
done